### DISCUSSION TITLE
routing and "perspective"

### DISCUSSION THREAD

**Discussion by chris-kruining** [podusowski/walkers#50] [Ideas]
I am hoping to make a navigation app for my car centerconsole project. I have 'simple' map rendering working in my egui app. but that is still a far cry from actual navigation. now what I think I need is routing (which I believe open streep maps offers) and the ability to render the map at an angle, like every other navigation app that is out there. I would like some advice on how to achieve this, I didn't straight up want to make feature requests as this might very well be doable without needing to put more features into walkers itself

routing: would you consider adding this? Or should I figure out how to fetch and render this on my own.

rendering on an angle/perspective: as far as I am aware egui is rendering to polygons underneath the hood, so is it possible to render a walkers map to a buffer/texture. display that on a quad, and angle that?

sorry in advance if I am asking dumb questions, still fairly new to rust, and super new to egui

**Comment by chris-kruining** [comment]
basically that image on the right is what I want to achieve
https://wiki.openstreetmap.org/wiki/Routing

albeit with more fancy graphics, hopefully

**Comment by podusowski** [comment]
HI!
That is a very cool project! :) Unfortunately, features you described are rather big, thus I am not sure whether I can spare time to implement them by myself.

You can implement routing (or at least the final part - drawing the lines) as walkers plugin. I'm sure algorithms for finding a path are there somewhere, so it shouldn't be a problem, but the real challenge is that you would need to have an access to the OSM (or other provider's) raw data - [`Planet.osm`](https://wiki.openstreetmap.org/wiki/Planet.osm). Since it is huge, you would need to either slice it by yourself, or use some third party services.

As for rendering, yeah, it's pretty much what you described, it would be best for walkers to draw the map on a texture, and then render it on some egui's surface. It seem like egui is already capable of providing a [low level access](https://github.com/emilk/egui/blob/master/crates/egui_demo_app/src/apps/custom3d_glow.rs) api for that.

**Comment by chris-kruining** [comment]
@podusowski do you have any clue what shader language glow uses? I have tried both glsl and wgsl editors and they both give syntax errors on the shaders used by the triangle example of egui.

**Reply by podusowski** [reply]
Sorry, no idea. I've never used it.
